src/commands/test/lib/included-files-parser.ts

```diff
@@ -1,8 +1,36 @@
-import { IFileDescriptionJson } from "./test-manifest-reader";
+import { IFileDescriptionJson, ITestCloudManifestJson } from "./test-manifest-reader";
 import * as path from "path";
+import * as pfs from "../../../util/misc/promisfied-fs";
+import { validXmlFile } from "./xml-util";
+import { out } from "../../../util/interaction";
+import _ = require("lodash");
 
 const invalidCharactersRegexp = /['"!#$%&+^<=>`|]/;
 
+export async function copyIncludedFiles(manifest: ITestCloudManifestJson, include: string[], rootDir: string): Promise<ITestCloudManifestJson> {
+  if (!include) {
+    return manifest;
+  }
+
+  const includedFiles = this.parseIncludedFiles(include, rootDir);
+
+  for (let i = 0; i < includedFiles.length; i++) {
+    const includedFile = includedFiles[i];
+    const copyTarget = path.join(path.dirname(rootDir), includedFile.targetPath);
+
+    if (_.endsWith(copyTarget, ".dll.config") && (manifest.files.indexOf(includedFile.targetPath.slice(0, -7)) > -1
```

Comment:
either work on copyTarget or includedFile.targetPath, not both. Preferable includedFile.targetPath since the logic is deciding if it should be copied at all)  

The if statement is really hard to read and I'm not sure it is working correctly. copyTarget file will not exist, its first created/copied a few files down. (await pfs.cp(includedFile.sourcePath, copyTarget))

I think something like this is more readable: 
```
for (let i = 0; i < includedFiles.length; i++) {
  const includedFile = includedFiles[i];
 if (_.endsWith(includedFile.targetPath, ".dll.config") {
   const assemblyName = includedFile.targetPath.slice(0, -7);
   const hasCorrespondingAssembly = manifest.files.indexOf(assemblyName) > -1 || include.indexOf(assemblyName);
  if(hasCorrespondingAssembly && !validXmlFile(includedFile.targetPath) {
      out.text(`Warning: The XML config file ${copyTarget} was not a valid XML file. This file will not be uploaded.`);
      continue;
   }
  }
   const copyTarget = path.join(path.dirname(rootDir), includedFile.targetPath);   
   await pfs.cp(includedFile.sourcePath, copyTarget);
   manifest.files.push(includedFile.targetPath);
 }
```
Sorry for the formatting and compile error, it was typed in the browser, but hopefully understandable :-)

Reply:
Good call! I don't know why I was using `copyTarget` in the first place - changing.